### PR TITLE
Add fl-with-flower post

### PR DESCRIPTION
**Pull request by charlesbvll**
Add a blog post about using Hugging Face in conjunction with Flower to train a fine-tune a Transformer model in a Federated setting (with data distributed amongst clients).

### CODE REVIEW

**Review by julien-c — COMMENTED** [review]
WDYT @osanseviero?

File: fl-with-flower.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+```python
+def load_data():
+    """Load IMDB data (training and eval)"""
+    raw_datasets = load_dataset("imdb")
+    raw_datasets = raw_datasets.shuffle(seed=42)
+
+    # remove unnecessary data split
+    del raw_datasets["unsupervised"]
+
+    tokenizer = AutoTokenizer.from_pretrained(CHECKPOINT)
```
Comment (osanseviero) [inline fl-with-flower.md:31]: Would be good to have `CHECKPOINT` defined

File: fl-with-flower.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+## Federating the example
+
+The idea behind Federated Learning is to train a model between multiple clients and a server, without having to share any data. This is done by letting each client train the model locally on its data and send its parameters back to the server which then aggregates all the clients’ parameters together using a predefined strategy. This process is made very simple by using the [Flower](https://github.com/adap/flower) framework. If you want a more complete overview, be sure to check out this guide: [What is Federated Learning?](https://flower.dev/docs/tutorial/Flower-0-What-is-FL.html)
+
+### Creating the IMDBClient
+
+To federate our example to multiple clients, we first need to write our Flower client class (inheriting from `flwr.client.NumPyClient`). This is very easy, as our model is a standard `PyTorch` model:
+
+```python
+class IMDBClient(fl.client.NumPyClient):
```
Comment (osanseviero) [inline fl-with-flower.md:114]: here and in other cells I would add imports corresponding to the cell

File: fl-with-flower.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+            print("Training Finished.")
+            return self.get_parameters(config={}), len(trainloader), {}
+
+        def evaluate(self, parameters, config):
+            self.set_parameters(parameters)
+            loss, accuracy = test(net, testloader)
+            return float(loss), len(testloader), {"accuracy": float(accuracy)}
+```
+
+The `get_parameters` function allows the server to get the client's parameters, inversely, the `set_parameters` function allows the server to send its parameters to the client. Finally the `fit` function is to train the model locally for the client and the `evaluate` function is to test the model locally and return the relevant metrics.
```
Comment (osanseviero) [inline fl-with-flower.md:136]:
```suggestion
The `get_parameters` function lets the server get the client's parameters. Inversely, the `set_parameters` function allows the server to send its parameters to the client. Finally, the `fit` function trains the model locally for the client, and the `evaluate` function tests the model locally and returns the relevant metrics. 
```

File: fl-with-flower.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+We can now start client instances using:
+
+```python
+flwr.client.start_numpy_client(server_address="127.0.0.1:8080", 
+															 client=IMDBClient())
+```
+
+### Starting the server
+
+Once we have a way to instantiate clients, we need to create our server in order to aggregate the results. Using Flower, this can be done very easily, by first choosing a strategy and then using the `flwr.server.start_server` function:
```
Comment (osanseviero) [inline fl-with-flower.md:147]:
```suggestion
Once we have a way to instantiate clients, we need to create our server in order to aggregate the results. Using Flower, this can be done very easily by first choosing a strategy and then using the `flwr.server.start_server` function:
```

**Comment by charlesbvll** [comment]
@osanseviero Thanks for all the feedback! I tried to address all your comments and suggestions. I just had a question about the notebook, currently, I just saved it from Colab, would you prefer if I strip out the metadata from it?

**Comment by osanseviero** [comment]
> would you prefer if I strip out the metadata from it?

I think as is is good!

File: fl-with-flower.md [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+    tokenizer = AutoTokenizer.from_pretrained(CHECKPOINT)
+
+    def tokenize_function(examples):
+        return tokenizer(examples["text"], truncation=True)
+
+    # random 100 samples
+    population = random.sample(range(len(raw_datasets["train"])), 100)
+
+    tokenized_datasets = raw_datasets.map(tokenize_function, batched=True)
+    tokenized_datasets["train"] = tokenized_datasets["train"].select(population)
```
Comment (osanseviero) [inline fl-with-flower.md:63]: Maybe add a comment on why we are selecting just a couple of samples

File: fl-with-flower.md [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+</a>
+
+This tutorial will show how to leverage Hugging Face to federate the training of language models over multiple clients using [Flower](https://flower.dev/). More specifically, we will fine-tune a pre-trained Transformer model (distilBERT) for sequence classification over a dataset of IMDB ratings. The end goal is to detect if a movie rating is positive or negative.
+
+## Dependencies
+
+To follow along this tutorial you will need to install the following packages: `datasets`, `flwr`, `torch`, and `transformers`. This can be done using `pip`:
+
+```sh
+pip install datasets flwr torch transformers
```
Comment (osanseviero) [inline fl-with-flower.md:25]: Maybe mention to use `flwr["simulation"]` if in colab

File: fl-with-flower.md [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+fl.client.start_numpy_client(server_address="127.0.0.1:8080", 
+															 client=IMDBClient())
+```
+
+### Starting the server
+
+Now that we have a way to instantiate clients, we need to create our server in order to aggregate the results. Using Flower, this can be done very easily by first choosing a strategy (here, we are using `FedAvg`, which will define the global weights as the average of all the clients' weights at each round) and then using the `flwr.server.start_server` function:
+
+```python
+import flwr as fl
```
Comment (osanseviero) [inline fl-with-flower.md:190]: If this is already imported once above, no need to import every single time, just the first time

File: fl-with-flower.md [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+
+    return trainloader, testloader
+```
+
+### Training and testing the model
+
+Once we have a way of creating our trainloader and testloader, we can take care of the training and testing. This is very similar to any `PyTorch` training or testing loop:
+
+```python
+import torch
```
Comment (osanseviero) [inline fl-with-flower.md:89]: No need to import torch in every cell, just the first time it's used

File: fl-with-flower.md [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+)
+```
+
+The `weighted_average` function is there to provide a way to aggregate the metrics distributed amongst the clients (basically this allows us to display a nice average accuracy and loss for every round).
+
+## Putting everything together
+
+If you want to check out everything put together, you should check out the code example we wrote for the Flower repo: [https://github.com/adap/flower/tree/main/examples/quickstart_huggingface](https://github.com/adap/flower/tree/main/examples/quickstart_huggingface). 
+
+A notebook is also available here: [https://colab.research.google.com/github/huggingface/blog/blob/main/notebooks/fl-with-flower.ipynb](https://colab.research.google.com/github/huggingface/blog/blob/main/notebooks/fl-with-flower.ipynb) but instead of running on multiple separate clients it utilizes the simulation functionality of Flower in order to emulate a federated setting inside Google Colab (this also means that instead of calling `start_server` we will call `start_simulation`, and that a few other modifications are needed).
```
Comment (osanseviero) [inline fl-with-flower.md:220]: Would maybe put this closer to the beginning

**Review by osanseviero — APPROVED** [review]
Cool!

File: fl-with-flower.md [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+    return trainloader, testloader
+```
+
+### Training and testing the model
+
+Once we have a way of creating our trainloader and testloader, we can take care of the training and testing. This is very similar to any `PyTorch` training or testing loop:
+
+```python
+import torch
+from datasets import load_metric
```
Comment (lhoestq) [inline fl-with-flower.md:90]: metrics are deprecated in `datasets`, you can use `evaluate` instead:

```suggestion
from evaluate import load as load_metric
```

**Comment by charlesbvll** [comment]
Hi @osanseviero! Should I change the date to Monday?

**Comment by osanseviero** [comment]
Yes. The blog can be merged already, but I think on Monday it will likely do better. Is that ok with you?

**Comment by lhoestq** [comment]
don't forget to update the notebook with the latest code changes ;)

**Comment by osanseviero** [comment]
Happy to merge as soon as the conflict is resolved and the notebook has the latest updates

**Comment by charlesbvll** [comment]
@osanseviero Thanks for the heads up! I think it can now be merged

**Comment by osanseviero** [comment]
https://huggingface.co/blog/fl-with-flower is out! :hugs: :tulip:

**Comment by osanseviero** [comment]
Hey there! Were there any communications on this? We would be delighted to amplify them and share them :hugs:

**Comment by charlesbvll** [comment]
Hey @osanseviero, that would be great! We're planning on releasing our communications at around 12 GMT on the 29th (sorry about that delay ^^), I'll send you the links then, thanks!

**Comment by charlesbvll** [comment]
@osanseviero Here are the posts: https://www.linkedin.com/posts/flwrlabs_flwr-flower-flowerlabs-activity-7046783043767848962-o3Ks?utm_source=share&utm_medium=member_desktop, https://twitter.com/flwrlabs/status/1641050406133329920! We might post another one on reddit, I will send you the link subsequently

**Comment by julien-c** [comment]
kudos on shipping this @charlesbvll !

**Comment by osanseviero** [comment]
Lots of success presenting at the OS Meetup! :fire:
